### DISCUSSION TITLE
State and business logic: All in the machine or some in the component?

### DISCUSSION THREAD

**Discussion by mtighe** [statelyai/xstate#2140] [Q&A]
Working on my first use of XState to clean up logic in a wizard-ish page. I have several hooks to pull in a few bits of data, which will update dynamically. I'm watching for these values to change in `useEffect`s, and sending events to the machine when I see changes that meet certain conditions. I'm also displaying some of this data, and I simply use the data returned from the hook and do not store it the machine context. Essentially, then, the machine manages which "state" I'm in (which is tied to what step gets displayed), but there is still logic and "state" stored outside of the machine, and I'm wondering if I should consolidate everything into the machine. I will attempt a very simplified example

```
const MyComponent = () => {
  const [current, send] = useMachine(myMachine);
  const data = useChangingData();
  
  useEffect(() => {
    if (data.foo > 5) send({ type: "DATA_CONDITION_MET" });
  }, [send, data]);
  
  return current.value === "initial" ? <div>Waiting</div> : <div>Foo: {data.foo}</div>;
}
```

Of course, in reality there are several of these and a bunch of timeouts and things that the machine is handling with many more states. Would it be better if I moved more of the logic into the machine, so I would get

```
const MyComponent = () => {
  const [current, send] = useMachine(myMachine);
  const data = useChangingData();
  useEffect(() => send({ type: "DATA_CHANGED", data }), [send, data]);
  
  const { context, value } = current;
  return value === "initial" ? <div>Waiting</div> : <div>Foo: {context.data.foo}</div>;
}
```

In this case I would store `data` in the machine context, and add a guard to the transition to the next state to check the `.foo` value. I'm also getting the data value to render from the machine itself rather than the original value from the hook (I could even pull that out into a separate hook to hide it entirely).

Which is more idiomatic? Should I try to push as much as possible into the machine? I feel like after typing this all out, the answer is "yes I should", but would like to hear any other thoughts.

**Comment by phuoccss** [comment]
Personally, I prefer having all business logic in the machine as it's more organized and easier to read and understand. Then components only handle UI rendering and sending events.
